subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultTypeAnnotationMetadataStore.java

```diff
@@ -349,6 +361,18 @@ private void processMethodAnnotations(Method method, Map<String, PropertyAnnotat
         }
     }
 
+    private static void validateSetterForMutableType(Method method, ValidationErrorsBuilder errorsBuilder, PropertyAccessorType accessorType) {
+        Class<?> setterType = getSetterType(method);
+        if (MUTABLE_NON_FINAL_TYPES.contains(setterType)) {
+            errorsBuilder.recordError(String.format("Support for property setters for type '%s' will be removed in a future release. Please consider making '%s' final.", setterType.getName(), accessorType.propertyNameFor(method)));
+        }
+    }
+
+    private static Class<?> getSetterType(Method method) {
+        Parameter firstParameter = method.getParameters()[0];
```

Comment:
is there a better way to get the type of the setter?

Reply:
You can use this: https://github.com/gradle/gradle/blob/eb88c3984da39864e0896abc2bcf655a30282528/subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java#L50-L56

The `PropertyAccessorType` should already be determined when you get here.

Reply:
yes, indeed, I was already using `propertyNameFor`, thanks!

Reply:
Please let me know if you think the end result is better than before.